File: functions/src/services/openai.service.ts
/**
 * OpenAI Service
 * Handles AI-powered card generation using GPT-3.5 Turbo
 */

import OpenAI from 'openai';
import { Card, CardType, UserProfile } from '../types';
import { CONSTANTS } from '../config/constants';
import { validateCard, fixInvalidUrls } from '../utils/validation.utils';
import { StockPriceService } from './stock-price.service';
import * as functions from 'firebase-functions';

export class OpenAIService {
  private client: OpenAI;
  private stockPriceService: StockPriceService;

  constructor(apiKey: string, stockPriceService: StockPriceService) {
    this.client = new OpenAI({ apiKey });
    this.stockPriceService = stockPriceService;
  }

  /**
   * Generate cards using OpenAI
   */
  async generateCards(
    profile: UserProfile,
    type: CardType,
    count: number,
    excludeList: string[],
    rotationStrategy: string,
    promptBuilder: (profile: UserProfile, type: CardType, count: number, excludeList: string[], rotationStrategy: string) => string
  ): Promise<Card[]> {
    const prompt = promptBuilder(profile, type, count, excludeList, rotationStrategy);

    try {
      console.log(`🤖 Calling OpenAI API...`);

      const completion = await this.client.chat.completions.create({
        model: CONSTANTS.OPENAI.MODEL,
        messages: [
          {
            role: 'system',
            content: 'You are a financial education assistant that generates personalized, beginner-friendly investment recommendations in JSON format.',
          },
          {
            role: 'user',
            content: prompt,
          },
        ],
        max_tokens: CONSTANTS.OPENAI.MAX_TOKENS,
        temperature: CONSTANTS.OPENAI.TEMPERATURE,
      });

      const content = completion.choices[0]?.message?.content || '';

      // Clean JSON response
      let cleanJSON = content.trim();

      // Remove markdown code blocks
      if (cleanJSON.startsWith('```json')) {
        cleanJSON = cleanJSON.replace(/```json\n?/g, '').replace(/```\n?/g, '');
      } else if (cleanJSON.startsWith('```')) {
        cleanJSON = cleanJSON.replace(/```\n?/g, '');
      }

      cleanJSON = cleanJSON.trim();

      let cards = JSON.parse(cleanJSON);

      if (!Array.isArray(cards) || cards.length === 0) {
        throw new Error('Invalid response: expected array of cards');
      }

      // If stocks, fetch real prices
      if (type === 'stock') {
        cards = await this.enrichWithRealPrices(cards);
      }

      // Validate cards
      const validCards = this.validateCards(cards, type);

      if (validCards.length === 0) {
        throw new functions.https.HttpsError(
          'internal',
          'No valid cards generated. Please try again.'
        );
      }

      // Fix invalid URLs
      fixInvalidUrls(validCards);

      console.log(`✅ Generated ${validCards.length} valid ${type} cards`);

      return validCards;
    } catch (error: any) {
      console.error('❌ OpenAI API error:', error);
      throw new functions.https.HttpsError('internal', 'Failed to generate cards', error.message);
    }
  }

  /**
   * Enrich stock cards with real-time prices
   */
  private async enrichWithRealPrices(cards: any[]): Promise<any[]> {
    console.log(`💰 Fetching real-time prices for ${cards.length} stocks...`);

    const tickers = cards.map((card: any) => card.ticker).filter(Boolean);
    const priceMap = await this.stockPriceService.fetchStockPrices(tickers);

    // Enrich cards with real prices
    cards.forEach((card: any) => {
      const realPrice = priceMap.get(card.ticker);
      if (realPrice) {
        card.price = realPrice.price;
        card.change = realPrice.changePercent;
        console.log(`✅ ${card.ticker}: ${realPrice.price} (${realPrice.changePercent})`);
      } else {
        // Keep AI estimate but log warning
        console.warn(`⚠️ Using estimate for ${card.ticker}`);
      }
    });

    return cards;
  }

  /**
   * Validate generated cards
   */
  private validateCards(cards: any[], type: CardType): Card[] {
    const validCards: Card[] = [];

    cards.forEach((card: any, index: number) => {
      const validation = validateCard(card, type);

      if (validation.valid) {
        validCards.push(card);

        if (validation.warnings.length > 0) {
          console.warn(`⚠️ Card ${index} warnings:`, validation.warnings);
        }
      } else {
        console.error(`❌ Card ${index} invalid:`, validation.errors);
      }
    });

    console.log(`✅ Valid cards: ${validCards.length}/${cards.length}`);

    return validCards;
  }
}

File: TickerCodebase/ViewModels/OpenAIService.swift
import Foundation
import OpenAI

class OpenAIService {
    static let shared = OpenAIService()
    
    private var client: OpenAI
    
    private init() {
        self.client = OpenAI(apiToken: AppConstants.openAIKey)
    }
    
    func generateInvestments(for profile: UserProfile, type: InvestmentType, count: Int = 5) async throws -> [Investment] {
        let prompt = buildPrompt(for: profile, type: type, count: count)
        
        // Create messages using correct TextContent & Content helpers
        let systemMessage = ChatQuery.ChatCompletionMessageParam.system(
            .init(content: .textContent("You are a financial education assistant that generates personalized, beginner-friendly investment recommendations in JSON format."))
        )
        let userMessage = ChatQuery.ChatCompletionMessageParam.user(
            .init(content: .string(prompt))
        )
        
        let query = ChatQuery(
            messages: [systemMessage, userMessage],
            model: .gpt4
        )
        
        let result = try await client.chats(query: query)
        
        guard let firstChoice = result.choices.first,
              let content = firstChoice.message.content else {
            throw OpenAIError.invalidResponse
        }
        
        return try parseInvestments(from: content, type: type)
    }
    
    private func buildPrompt(for profile: UserProfile, type: InvestmentType, count: Int) -> String {
        let typeString = type == .stock ? "stocks" : "business ideas"
        
        return """
        Generate \(count) personalized \(typeString) recommendations for a beginner investor with this profile:
        
        \(profile.openAIPrompt)
        
        Return ONLY a valid JSON array with no additional text. Each item must follow this exact structure:
        
        \(type == .stock ? stockJSONExample : ideaJSONExample)
        
        Requirements:
        - Write in simple, beginner-friendly language
        - Keep "goodReasons" and "concerns" to 3-4 items each
        - Make taglines catchy and under 15 words
        - Include real, clickable URLs for sources and getStarted resources
        - For stocks: use real current prices and ticker symbols
        - For ideas: provide realistic investment ranges
        - All text should educate and inform, not hype
        
        Return ONLY the JSON array, nothing else.
        """
    }
    
    private var stockJSONExample: String {
        """
        [
          {
            "type": "stock",
            "title": "NVIDIA",
            "ticker": "NVDA",
            "price": "$875.32",
            "change": "+2.4%",
            "tagline": "The company making chips that power AI",
            "simpleExplainer": "NVIDIA makes the computer chips that run ChatGPT, self-driving cars, and most AI technology.",
            "whatToExpect": "The stock price can swing up or down quickly based on AI news. You're betting that AI keeps growing.",
            "goodReasons": [
              "They make most of the AI chips in the world",
              "Big tech companies like Microsoft buy from them",
              "AI is still in early stages"
            ],
            "concerns": [
              "The stock is expensive right now",
              "Other companies are trying to compete",
              "If AI hype slows down, stock could drop"
            ],
            "timeline": "3-5 years",
            "riskLevel": "Medium-High",
            "beginnerTip": "This is a growth stock - you buy hoping the company gets bigger, not for steady income.",
            "sources": [
              {"name": "Yahoo Finance", "url": "https://finance.yahoo.com/quote/NVDA"},
              {"name": "Seeking Alpha", "url": "https://seekingalpha.com/symbol/NVDA"}
            ],
            "getStarted": [
              {"name": "Robinhood", "description": "Easy app to buy stocks", "url": "https://robinhood.com"},
              {"name": "Fidelity", "description": "Full-service broker", "url": "https://fidelity.com"}
            ]
          }
        ]
        """
    }
    
    private var ideaJSONExample: String {
        """
        [
          {
            "type": "idea",
            "title": "AI Tools for Doctor Offices",
            "category": "Software Business",
            "investment": "$50K - $150K",
            "tagline": "Build automation software that saves doctors time",
            "simpleExplainer": "Doctors spend hours on paperwork. You could build AI tools that automatically write up patient notes.",
            "whatToExpect": "Takes 12-18 months to get your first paying customers. Sales are slow because doctors are busy.",
            "goodReasons": [
              "Healthcare is always growing",
              "Once customers pay, they keep paying monthly",
              "Good profit margins once you have customers"
            ],
            "concerns": [
              "Strict rules about patient privacy (HIPAA)",
              "Doctors take 3-6 months to decide to buy",
              "You need to know how to code or hire developers"
            ],
            "timeline": "18-24 months to make profit",
            "riskLevel": "Medium",
            "beginnerTip": "This is called SaaS - customers pay you monthly. It takes time to start but can be very profitable.",
            "sources": [
              {"name": "CB Insights", "url": "https://cbinsights.com/healthcare-ai"},
              {"name": "Grand View Research", "url": "https://grandviewresearch.com"}
            ],
            "getStarted": [
              {"name": "Bubble.io", "description": "Build apps without coding", "url": "https://bubble.io"},
              {"name": "Stripe", "description": "Accept monthly payments", "url": "https://stripe.com"},
              {"name": "Canva", "description": "Design marketing materials", "url": "https://canva.com"}
            ]
          }
        ]
        """
    }
    
    private func parseInvestments(from json: String, type: InvestmentType) throws -> [Investment] {
        // Clean up potential markdown code blocks
        let cleanJSON = json
            .replacingOccurrences(of: "```json", with: "")
            .replacingOccurrences(of: "```", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard let data = cleanJSON.data(using: .utf8) else {
            throw OpenAIError.invalidJSON
        }
        
        let decoder = JSONDecoder()
        let investments = try decoder.decode([Investment].self, from: data)
        return investments
    }
}

enum OpenAIError: LocalizedError {
    case invalidResponse
    case invalidJSON
    case apiError(String)
    
    var errorDescription: String? {
        switch self {
        case .invalidResponse: return "Could not get response from AI"
        case .invalidJSON: return "Could not parse AI response"
        case .apiError(let message): return message
        }
    }
}
